src/storage/v2/disk/storage.cpp

```diff
@@ -0,0 +1,1601 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include <limits>
+#include <optional>
+#include <stdexcept>
+#include <vector>
+
+#include <rocksdb/comparator.h>
+#include <rocksdb/db.h>
+#include <rocksdb/slice.h>
+
+#include <rocksdb/options.h>
+#include <rocksdb/utilities/transaction.h>
+#include <rocksdb/utilities/transaction_db.h>
+
+#include "kvstore/kvstore.hpp"
+#include "spdlog/spdlog.h"
+#include "storage/v2/constraints/unique_constraints.hpp"
+#include "storage/v2/disk/rocksdb_storage.hpp"
+#include "storage/v2/disk/storage.hpp"
+#include "storage/v2/disk/unique_constraints.hpp"
+#include "storage/v2/id_types.hpp"
+#include "storage/v2/mvcc.hpp"
+#include "storage/v2/property_store.hpp"
+#include "storage/v2/property_value.hpp"
+#include "storage/v2/result.hpp"
+#include "storage/v2/storage.hpp"
+#include "storage/v2/storage_error.hpp"
+#include "storage/v2/transaction.hpp"
+#include "storage/v2/vertex_accessor.hpp"
+#include "storage/v2/view.hpp"
+#include "utils/disk_utils.hpp"
+#include "utils/exceptions.hpp"
+#include "utils/file.hpp"
+#include "utils/memory_tracker.hpp"
+#include "utils/message.hpp"
+#include "utils/on_scope_exit.hpp"
+#include "utils/readable_size.hpp"
+#include "utils/rocksdb_serialization.hpp"
+#include "utils/skip_list.hpp"
+#include "utils/stat.hpp"
+#include "utils/string.hpp"
+
+namespace memgraph::storage {
+
+using OOMExceptionEnabler = utils::MemoryTracker::OutOfMemoryExceptionEnabler;
+
+namespace {
+
+constexpr const char *vertexHandle = "vertex";
+constexpr const char *edgeHandle = "edge";
+constexpr const char *defaultHandle = "default";
+constexpr const char *lastTransactionStartTimeStamp = "last_transaction_start_timestamp";
+constexpr const char *label_index_str = "label_index";
+constexpr const char *label_property_index_str = "label_property_index";
+constexpr const char *existence_constraints_str = "existence_constraints";
+constexpr const char *unique_constraints_str = "unique_constraints";
+
+bool VertexExistsInCache(const utils::SkipList<Vertex>::Accessor &accessor, Gid gid) {
+  return accessor.find(gid) != accessor.end();
+}
+
+bool VertexHasLabel(const Vertex &vertex, LabelId label, Transaction *transaction, View view) {
+  bool deleted = false;
+  bool has_label = false;
+  Delta *delta = nullptr;
+  {
+    std::lock_guard<utils::SpinLock> guard(vertex.lock);
+    deleted = vertex.deleted;
+    has_label = std::find(vertex.labels.begin(), vertex.labels.end(), label) != vertex.labels.end();
+    delta = vertex.delta;
+  }
+  ApplyDeltasForRead(transaction, delta, view, [&deleted, &has_label, label](const Delta &delta) {
+    switch (delta.action) {
+      case Delta::Action::REMOVE_LABEL: {
+        if (delta.label == label) {
+          MG_ASSERT(has_label, "Invalid database state!");
+          has_label = false;
+        }
+        break;
+      }
```

Comment:
Todo: we need to check this if it is correct behaviour: https://memgraph.slack.com/archives/CS5827FK3/p1687530832459949